content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,157 @@
+
+![6_edits](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/6_edits.png)
+
+Each of these features works very differently from one another. The mental model of the user and the requirements are different for each. **For autocomplete, speed is key.** You want it to be very fast and accurate enough so that it doesn’t interrupt the programmer’s flow. On the other hand, **for chat, accuracy and the quality of the responses may be more important**.
+
+![7_tradeoff](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/7_tradeoff.png)
+
+When generating tests, there may be additional requirements. For example, if you’re looking at a file and want to add tests to it, you need to quickly detect if a test file for that file already exists. So, when we say “coding assistant,” it might sound simple, but it’s actually a product of products—multiple products put together to create a comprehensive experience.
+
+In this post, we're going to take a look at how context fetching works for each particular feature - chat, autocomplete, generating tests, and editing code.
+
+---
+
+## How context fetching works for chat
+
+![8_chat](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/8_chat.png)
+
+Chat allows you to ask questions about your codebase and generate code that's tailored to work within your codebase. For this, we want a few things:
+
+1. **Conversation history:** In a given chat session, we record previous messages as they may contain relevant context for the user's next request.
+2. **Code search:** We fetch the most relevant code snippets related to the user's query from the codebase using search, similar to how a human dev might search for these code snippets.
+3. **User control:** Users should have the ability to mention specific files and provide those directly to the model with an option to reference external sources like Slack threads or Notion documents to enrich the context further.
+
+![9_context](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/9_context.png)
+
+For code search, we use two techniques for surfacing relevant snippets:
+
+1. **AI-enhanced keyword search:** We ask an LLM a question like, "Given this user query, what keywords should be used to search for relevant parts of the codebase?" We use these keywords to query a keyword search index we've created over the codebase. This index returns a set of results, ordered by Sourcegraph's code-aware ranking mechanism, and we select the top results subject to the constraint of the LLM's context window.
```

Comment:
Maybe we can just call this `Keyword Search`, and instead of saying "We ask the LLM...", we can emphasize all the different thoughtful transformations we do (some of which are just from 'traditional' search). So something like...

"We first perform a 'query understanding' step, which interprets the user's query and crafts a set of code searches to perform. This step pulls out important entities like file paths and symbols, as well as key terms and their variants. We use both traditional search techniques, and a lightweight LLM to process queries."